postrelease-vip.php

```diff
@@ -2,567 +2,680 @@
-    update_option('prx_template_post_id', $template_post_id);
-}
-
-
-function postrelease_template_redirect() {
-	global $wp_query;
-
-	// functions that don't require security check
-	if(isset($wp_query->query_vars['prx'])) { //prx is a URL parameter
-		$function = $wp_query->query_vars['prx'];	
-		
-		if ($function == 'ad') {	// redirect to full ad page		
-			if(get_option('prx_plugin_activated', 0) == 1) {
-
-				//redirect to template page
-				$template_post_id = get_option('prx_template_post_id', 1);
-
-				if ( ! $template_post_id )
-					return;
-
-				$template_post_url = get_permalink($template_post_id);
-
-				$query_struct = array();
-				parse_str($_SERVER['QUERY_STRING'], $query_struct);
-				$template_post_url = add_query_arg( $query_struct, $template_post_url ); //keeping URL parameters when we redirect (prx_t and prx_rk need to stay in the URL)				
-				$template_post_url = add_query_arg( 'prx', 'page', $template_post_url ); //prx=page so that we do not get redirected anymore
-				wp_redirect( $template_post_url );
-				
-				exit;
+		/**
+		 * Appends the PostRelease javascript to header
+		 *
+		 * @return void
+		 */
+		public function postrelease_wp_enqueue_scripts() {
+			wp_enqueue_script( 'postrelease', $this->postrelease_js_url, [], null, false );
+		}
+
+		/**
+		 * Register postrelease post type
+		 *
+		 * @return void
+		 */
+		public function postrelease_init() {
+			$args = [
+				'public' => false || $this->is_dev,
+				'publicly_queryable' => true,
+				'label' => 'Sponsored Post',
+				'exclude_from_search' => true,
+				'can_export' => false,
+				'supports' => [
+					'title',
+					'editor',
+					'author',
+					'thumbnail',
+					'excerpt',
+				],
+			];
+			register_post_type( $this->post_type, $args );
+		}
+
+		/**
+		 * Function when plugin is activated.
+		 * In WP VIP, the activate hook is not supported.
+		 * The plugin is activated when the user goes to the dashboard and
+		 * adds his publication to the PostRelease network.
+		 *
+		 * @return void
+		 */
+		public function postrelease_activate() {
+			if ( ! function_exists( 'wpcom_is_vip' ) ) {
+				flush_rewrite_rules();
+			} else {
+				do_action( 'postrelease_activate' );
 			}
 		}
-	}	
-}
 
-/**
- * Security check - validate key 
- * @param string $key security MD5 encoded
- */
-function postrelease_is_key_valid($key) {
-	if( false != get_option('prx_plugin_key', false) ) { //there is a key saved here
-		$md5_key = md5(get_option('prx_plugin_key'));
-		if(strcmp($md5_key,$key) === 0) {
+		/**
+		 * Reset changes caused by the plugin upon deactivation
+		 *
+		 * @return void
+		 */
+		public function postrelease_deactivate() {
+			// delete all sponsored posts
+			$sponsored_posts_array = get_pages( [ 'post_type' => $this->post_type ] );
+			if ( ! empty( $sponsored_posts_array ) ) {
+				foreach ( $sponsored_posts_array as $sponsored_post ) {
+					wp_delete_post( $sponsored_post->ID, true );
+				}
+			}
```

Comment:
The calling of the deactivate method is somewhat deceiving. This method is called from the `enable_plugin` method, which is just checking an option and then calling `$this->deactivate()`, which really isn't a WP deactivation in this case. There is a deactivation hook registered, but it calls the `notify_deactivate` method instead, which is just sending off a post request to the postrelease server.